File: problem9.cpp
#include <iostream>
#include <vector>

std::vector<int> solution()
{
  for ( int a = 1; a < 1000; ++a )
  {
    for ( int b = a; b < 1000; ++b )
    {
      for ( int c = b; c < 1000; ++c )
      {
        if ( a + b + c == 1000 && a*a + b*b == c*c )
        {
          return { a, b, c };
        }
      }
    }
  }
  return {-1,-1,-1};
}

int main()
{
  auto r = solution();
  std::cout << "a: " << r[0] << " b: " << r[1] << " c: " << r[2] << std::endl;
  std::cout << "product: " << r[0] * r[1] * r[2] << std::endl;
}

File: ProjectEuler/problem9.hs
-- These problems from projecteuler.net
--
-- Problem 9
-- A Pythagorean triplet is a set of three natural numbers, a < b < c, for which,
--
-- a^2 + b^2 = c^2
-- For example, 3^2 + 4^2 = 9 + 16 = 25 = 5^2.
--
-- There exists exactly one Pythagorean triplet for which a + b + c = 1000.
-- Find the product abc.

-- NOT TESTED
{- TODO needs some (math based) optimization
pithag :: (Num a, Eq a, Ord a) => [a] -> [(a,a,a)]
pithag xs = [ (a,b,c) | a <- xs,
                        b <- xs,
                        c <- xs,
                a + b + c == 1000,
                a^2 + b^2 == c^2 ]

solution = head $ pithag [1..1000]
-}


